linkerd/app/inbound/src/http/server.rs

```diff
@@ -127,23 +127,23 @@ impl<T: Param<tls::ConditionalServerTls>> ExtractParam<errors::respond::EmitHead
 
 impl errors::HttpRescue<Error> for ServerRescue {
     fn rescue(&self, error: Error) -> Result<errors::SyntheticHttpResponse> {
-        let cause = errors::root_cause(&*error);
-        if cause.is::<crate::policy::DeniedUnauthorized>() {
+        if let Some(cause) = errors::cause_ref::<crate::policy::DeniedUnauthorized>(&*error) {
             return Ok(errors::SyntheticHttpResponse::permission_denied(cause));
         }
-        if cause.is::<crate::GatewayDomainInvalid>() {
+        if let Some(cause) = errors::cause_ref::<crate::GatewayDomainInvalid>(&*error) {
             return Ok(errors::SyntheticHttpResponse::not_found(cause));
         }
-        if cause.is::<crate::GatewayIdentityRequired>() {
+        if let Some(cause) = errors::cause_ref::<crate::GatewayIdentityRequired>(&*error) {
             return Ok(errors::SyntheticHttpResponse::unauthenticated(cause));
         }
-        if cause.is::<crate::GatewayLoop>() {
+        if let Some(cause) = errors::cause_ref::<crate::GatewayLoop>(&*error) {
             return Ok(errors::SyntheticHttpResponse::loop_detected(cause));
         }
-        if cause.is::<errors::FailFastError>() {
+        if let Some(cause) = errors::cause_ref::<errors::FailFastError>(&*error) {
             return Ok(errors::SyntheticHttpResponse::gateway_timeout(cause));
         }
```

Comment:
this does mean that we will now traverse the entire cause chain for _every_ separate `if` condition, rather than doing it once to find the root cause and then testing that error's type. OTTOH, this code will now handle cases where the error of a desired type is not the ultimate root cause...although i believe all of these error types are always roots.

i don't know if the performance implications of iterating over the cause chain multiple times is really worth worrying about, but i wanted to point it out — i think having a single correct implementation of this is worth more than not traversing a couple nested causes a few more times, in any case.

Reply:
yeah that was my thought process as well.